src/thread_pool_async.ml

```diff
@@ -2,83 +2,113 @@ open Base
-    let%bind () = In_thread.run ~thread (fun () -> destroy state) in
-    In_thread.run ~thread create
-  in
-  let run_once
-      : 'thread -> 'state -> ('result computation Or_error.t * 'state) Deferred.t
-    =
-   fun thread state ->
-    let%bind result =
-      In_thread.run ~thread (fun () -> Or_error.try_with (fun () -> f state))
+let execute {thread; state} work = In_thread.run ~thread (fun () -> work state)
+
+let with_worker ({reader; writer; _} as pool) ?(retries = 0) job =
+  let run_once worker =
+    let%bind result = try_with (fun () -> job worker) in
+    let%bind worker =
+      match result with
+      | Ok (Ok _) -> return worker
+      | Ok (Error _) | Error _ -> recreate_state pool worker
     in
-    match result with
-    | Ok (`Ok res) -> return (Ok (`Ok res), state)
-    | Ok `Attempt_retry ->
-        let%bind state = recreate ~thread state in
-        return (Ok `Attempt_retry, state)
-    | Error _ as e ->
-        let%bind state = recreate ~thread state in
-        return (e, state)
+    return (result, worker)
+  in
+  let rec run_with_retry worker retries_left =
+    let%bind outcome, worker = run_once worker in
+    match outcome with
+    | Ok (Ok _) -> return (outcome, worker)
+    | Ok (Error _) -> (
+      match retries_left with
+      | n when n <= 0 -> return (outcome, worker)
+      | n -> run_with_retry worker (Int.pred n))
+    | Error _ -> return (outcome, worker)
   in
   match%bind Pipe.read reader with
-  | `Eof -> return @@ Or_error.errorf "Pool has been destroyed"
-  | `Ok (thread, state) ->
-      let rec run_with_retry state retries_left =
-        let%bind result, state = run_once thread state in
-        match result with
-        | Ok (`Ok result) -> return (Ok result, state)
-        | Ok `Attempt_retry -> (
-          match retries_left with
-          | n when n <= 0 -> return (Or_error.error_string "No more retries", state)
-          | n -> run_with_retry state (Int.pred n))
-        | Error err -> return (Error err, state)
-      in
-      let%bind result, state' = run_with_retry state retries in
-      Pipe.write_without_pushback writer (thread, state');
-      return result
+  | `Eof -> raise Pool_already_destroyed
+  | `Ok worker -> (
+      let%bind outcome, worker = run_with_retry worker retries in
+      Pipe.write_without_pushback writer worker;
+      match outcome with
+      | Ok outcome -> return outcome
+      | Error exn -> raise exn)
 
-let destroy {pool = reader, writer; threads; destroy; _} : unit Deferred.t =
-  let elements = List.init threads ~f:ignore in
-  let%bind () =
-    Deferred.List.iter elements ~how:`Parallel ~f:(fun () ->
-        match%bind Pipe.read reader with
-        | `Eof -> return ()
-        | `Ok (thread, state) -> In_thread.run ~thread (fun () -> destroy state))
-  in
-  return @@ Pipe.close writer
+let with' pool ?(retries = 0) work =
+  let job worker = execute worker work in
+  with_worker pool ~retries job
+
+let destroy = destroy_pool
```

Comment:
Maybe it would be nicer to call `destroy_pool` `destroy`, then there is no name duplication.

Reply:
That would conflict with the `destroy` function that destroys a piece of state.